File: fronted/src/api/agentTask.ts
import { get, post } from './http'

export type AgentTaskStatus = 'created' | 'queued' | 'running' | 'succeeded' | 'failed' | 'cancelled'

export interface AgentTask<TOutput = unknown> {
  id: number
  agent_key: string
  project_id?: number | null
  user_id?: number | null
  status: AgentTaskStatus
  priority: number
  input_payload: Record<string, unknown>
  result_payload?: {
    summary?: string
    output?: TOutput
    artifacts?: AgentArtifact[]
  } | null
  error_message?: string | null
  retry_count: number
  started_at?: string | null
  finished_at?: string | null
  created_at: string
  updated_at: string
}

export interface AgentTaskEvent {
  id: number
  task_id: number
  event_type: string
  message: string
  progress?: number | null
  payload?: Record<string, unknown> | null
  created_at: string
}

export interface AgentArtifact {
  id?: number
  task_id?: number
  name: string
  artifact_type: string
  storage_path: string
  mime_type?: string | null
  size_bytes?: number | null
  created_at?: string
}

export interface AgentTaskCreateRequest {
  agent_key: string
  project_id?: number | null
  user_id?: number | null
  priority?: number
  input_payload: Record<string, unknown>
}

export interface AgentTaskListResponse<TOutput = unknown> {
  items: AgentTask<TOutput>[]
  total: number
  page: number
  page_size: number
}

export interface AgentTaskQuery {
  agent_key?: string
  status?: AgentTaskStatus
  project_id?: number
  skip?: number
  limit?: number
}

export function createAgentTask<TOutput = unknown>(data: AgentTaskCreateRequest): Promise<AgentTask<TOutput>> {
  return post<AgentTask<TOutput>>('/api/v1/agent-tasks/', data)
}

export function getAgentTasks<TOutput = unknown>(params?: AgentTaskQuery): Promise<AgentTaskListResponse<TOutput>> {
  return get<AgentTaskListResponse<TOutput>>('/api/v1/agent-tasks/', params)
}

export function getAgentTask<TOutput = unknown>(taskId: number): Promise<AgentTask<TOutput>> {
  return get<AgentTask<TOutput>>(`/api/v1/agent-tasks/${taskId}`)
}

export function getAgentTaskEvents(taskId: number): Promise<AgentTaskEvent[]> {
  return get<AgentTaskEvent[]>(`/api/v1/agent-tasks/${taskId}/events`)
}

export function getAgentTaskArtifacts(taskId: number): Promise<AgentArtifact[]> {
  return get<AgentArtifact[]>(`/api/v1/agent-tasks/${taskId}/artifacts`)
}

export function retryAgentTask<TOutput = unknown>(taskId: number): Promise<AgentTask<TOutput>> {
  return post<AgentTask<TOutput>>(`/api/v1/agent-tasks/${taskId}/retry`)
}

export function cancelAgentTask<TOutput = unknown>(taskId: number): Promise<AgentTask<TOutput>> {
  return post<AgentTask<TOutput>>(`/api/v1/agent-tasks/${taskId}/cancel`)
}

File: backend/app/models/agent_task.py
from __future__ import annotations

from datetime import datetime
from enum import Enum

from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.types import JSON

from app.db.base_class import Base


class AgentTaskStatus(str, Enum):
    CREATED = "created"
    QUEUED = "queued"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    CANCELLED = "cancelled"


class AgentTask(Base):
    __tablename__ = "agent_tasks"

    id = Column(Integer, primary_key=True, index=True)
    agent_key = Column(String(100), nullable=False, index=True)
    project_id = Column(Integer, nullable=True, index=True)
    user_id = Column(Integer, nullable=True, index=True)
    status = Column(String(20), nullable=False, default=AgentTaskStatus.CREATED.value, index=True)
    priority = Column(Integer, nullable=False, default=0)
    input_payload = Column(JSON, nullable=False, default=dict)
    result_payload = Column(JSON, nullable=True)
    error_message = Column(Text, nullable=True)
    retry_count = Column(Integer, nullable=False, default=0)
    started_at = Column(DateTime, nullable=True)
    finished_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)


class AgentTaskEvent(Base):
    __tablename__ = "agent_task_events"

    id = Column(Integer, primary_key=True, index=True)
    task_id = Column(Integer, ForeignKey("agent_tasks.id"), nullable=False, index=True)
    event_type = Column(String(50), nullable=False, default="info")
    message = Column(Text, nullable=False)
    progress = Column(Integer, nullable=True)
    payload = Column(JSON, nullable=True)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)


class AgentArtifact(Base):
    __tablename__ = "agent_artifacts"

    id = Column(Integer, primary_key=True, index=True)
    task_id = Column(Integer, ForeignKey("agent_tasks.id"), nullable=False, index=True)
    name = Column(String(255), nullable=False)
    artifact_type = Column(String(50), nullable=False)
    storage_path = Column(String(1024), nullable=False)
    mime_type = Column(String(255), nullable=True)
    size_bytes = Column(Integer, nullable=True)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
